contracts/starknet/Authenticators/EthSig.cairo

```diff
@@ -0,0 +1,370 @@
+end
+
+# Adds a 2 bytes (16 bits) `prefix` to a 16 bytes (128 bits) `value`.
+func add_prefix128{range_check_ptr, bitwise_ptr : BitwiseBuiltin*}(value : felt, prefix : felt) -> (
+    result : felt, carry
+):
+    # Shift the prefix by 128 bits
+    let shifted_prefix = prefix * 2 ** 128
+    # `with_prefix` is now 18 bytes long
+    let with_prefix = shifted_prefix + value
+    # Create 2 bytes mask
+    let overflow_mask = 2 ** 16 - 1
+    # Extract the last two bytes of `with_prefix`
+    let (carry) = bitwise_and(with_prefix, overflow_mask)
+    # Compute the new number, right shift by 16
+    let result = (with_prefix - carry) / 2 ** 16
+    return (result, carry)
+end
+
+# Concatenates a 2 bytes long `prefix` and `input` to `output`.
+# `input_len` is the number of `Uint256` in `input`.
+func prepend_prefix_2bytes{range_check_ptr, bitwise_ptr : BitwiseBuiltin*}(
+    prefix : felt, output : Uint256*, input_len : felt, input : Uint256*
+):
+    if input_len == 0:
+        # Done, simlpy store the prefix in the `.high` part of the last Uin256, and
+        # make sure we left shift it by 28 (32 - 4)
+        assert output[0] = Uint256(0, prefix * 16 ** 28)
+        return ()
+    else:
+        let num = input[0]
+
+        let (w1, high_carry) = add_prefix128(num.high, prefix)
+        let (w0, low_carry) = add_prefix128(num.low, high_carry)
+
+        let res = Uint256(w0, w1)
+        assert output[0] = res
+
+        # Recurse, using the `low_carry` as `prefix`
+        prepend_prefix_2bytes(low_carry, &output[1], input_len - 1, &input[1])
+        return ()
+    end
+end
+
+# Computes the `keccak256` hash from an array of `Uint256`. Does NOT call `finalize_keccak`,
+# so the caller needs to make she calls `finalize_keccak` on the `keccak_ptr` once she's done
+# with it.
+func get_keccak_hash{range_check_ptr, bitwise_ptr : BitwiseBuiltin*, keccak_ptr : felt*}(
+    uint256_words_len : felt, uint256_words : Uint256*
+) -> (hash : Uint256):
+    let (hash) = keccak_uint256s_bigend{keccak_ptr=keccak_ptr}(uint256_words_len, uint256_words)
+
+    return (hash)
+end
+
+# Returns the number of digits needed to represent `num` in hexadecimal.
+# Similar to doing `len(hex(num)[2:])` in Python.
+# E.g.:
+# - `0x123` will return `3`
+# - `0x1` will return `1`
+# - `0xa3b1d4` will return `6`
+# Notice: Will not work for `0x0` (will return `0` for `0x0` instead of `1`).
+func get_base16_len{range_check_ptr}(num : Uint256) -> (res : felt):
+    let (is_eq) = uint256_eq(num, Uint256(0, 0))
+    if is_eq == 1:
+        return (0)
+    else:
+        # Divide by 16
+        let (divided, _) = uint256_unsigned_div_rem(num, Uint256(16, 0))
+
+        let (res_len) = get_base16_len(divided)
+        return (res_len + 1)
+    end
+end
+
+# Computes `base ** exp` where `base` and `exp` are both `felts` and returns the result as a `Uint256`.
+func u256_pow{range_check_ptr}(base : felt, exp : felt) -> (res : Uint256):
```

Comment:
any reason to not use [pow](https://github.com/starkware-libs/cairo-lang/blob/167b28bcd940fd25ea3816204fa882a0b0a49603/src/starkware/cairo/common/pow.cairo) from standard lib then just convert to Uint256 ?